middlewares/reqHost.go

```diff
@@ -0,0 +1,41 @@
+package middlewares
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/containous/traefik/types"
+)
+
+var reqHostKey struct{}
+
+// ReqHostMiddleware is the struct for the middleware that adds the CanonicalDomain of the requst Host into a context for later use.
+type ReqHostMiddleware struct{}
+
+// NewReqHostMiddleware is a middleware that adds the CanonicalDomain of the requst Host into a context for later use.
+func NewReqHostMiddleware() *ReqHostMiddleware {
+	return &ReqHostMiddleware{}
+}
+
+func (rhm *ReqHostMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	reqHost := r.Host
```

Comment:
You could maybe simplify like that 
```go
func (rh *ReqHostMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
	reqHost := types.CanonicalDomain(parseHost(r.Host))
	if next != nil {
		next.ServeHTTP(rw, r.WithContext(context.WithValue(r.Context(), reqHostKey, reqHost)))
	}
}

func parseHost(addr string) string {
	if !strings.Contains(addr, ":") {
		return addr
	}

	host, _, err := net.SplitHostPort(addr)
	if err != nil {
		return addr
	}
	return host
}
```

WDYT?

And you should be able to test `parseHost` function